File: 36.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#

# @lc code=start

from collections import defaultdict
from itertools import product


class Solution:
    def isValidSudoku(self, board):
        cols, rows, grid = defaultdict(set), defaultdict(set), defaultdict(set)

        for r, c in product(range(9), range(9)):
            tile = board[r][c]
            if tile == ".":
                continue
            if tile in (cols[c] | rows[r] | grid[r // 3, c // 3]):
                return False
            cols[c].add(tile)
            rows[r].add(tile)
            grid[r // 3, c // 3].add(tile)
        return True


# @lc code=end

File: 36.valid-sudoku.rs
/*
 * @lc app=leetcode id=36 lang=rust
 *
 * [36] Valid Sudoku
 */

// @lc code=start
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let (mut cols, mut rows, mut grids) = ([0u16; 9], [0u16; 9], [0u16; 9]);

        for r in 0..9 {
            for c in (0..9).filter(|&c| board[r][c] != '.') {

                let tile = 1 << (board[r][c] as u8 - b'1');
                let g = r - r % 3 + c / 3;

                if tile & (cols[c] | rows[r] | grids[g]) != 0 {
                    return false;
                }
                cols[c] |= tile;
                rows[r] |= tile;
                grids[g] |= tile;
            }
        }
        true
    }
}
// @lc code=end
